pandas/core/case_when.py

```diff
@@ -0,0 +1,217 @@
+            for condition, replacement in zip(conditions, replacements)
+        ]
+        if (default is not None) and isinstance(default, ABCNDFrame):
+            default = default.astype(common_dtype)
+    else:
+        common_dtype = common_dtype[0]
+    if not isinstance(default, ABCNDFrame):
+        cond_indices = [cond for cond in conditions if isinstance(cond, ABCNDFrame)]
+        replacement_indices = [
+            replacement
+            for replacement in replacements
+            if isinstance(replacement, ABCNDFrame)
+        ]
+        cond_length = None
+        if replacement_indices:
+            for left, right in zip_longest(
+                replacement_indices, replacement_indices[1:]
+            ):
+                if right is None:
+                    continue
+                if assert_index_equal(left.index, right.index, check_order=False):
+                    raise ValueError(
+                        "All replacement objects must have the same index."
+                    )
+        if cond_indices:
+            for left, right in zip_longest(cond_indices, cond_indices[1:]):
+                if right is None:
+                    continue
+                if assert_index_equal(left.index, right.index, check_order=False):
+                    raise ValueError("All condition objects must have the same index.")
+        if replacement_indices and cond_indices:
+            if assert_index_equal(
+                replacement_indices[0].index, cond_indices[0].index, check_order=False
+            ):
+                raise ValueError(
+                    "All replacement objects and condition objects "
+                    "should have the same index."
+                )
+        elif not cond_indices:
+            conditions = [
+                np.asanyarray(cond) if not hasattr(cond, "shape") else cond
+                for cond in conditions
+            ]
+            cond_length = {len(cond) for cond in conditions}
+            if len(cond_length) > 1:
+                raise ValueError("The boolean conditions should have the same length.")
+            cond_length = len(conditions[0])
+            if not is_scalar(default):
+                if len(default) != cond_length:
+                    raise ValueError(
+                        "length of `default` does not match the length "
+                        "of any of the conditions."
+                    )
+            if not replacement_indices:
+                for num, replacement in enumerate(replacements):
+                    if is_scalar(replacement):
+                        continue
+                    if not hasattr(replacement, "shape"):
+                        replacement = np.asanyarray(replacement)
+                    if len(replacement) != cond_length:
+                        raise ValueError(
+                            f"Length of condition{num} does not match "
+                            f"the length of replacement{num}; "
```

Comment:
Similar - this should only be enforced for non-Series objects.

Reply:
yes, it applies to the non-Series objects